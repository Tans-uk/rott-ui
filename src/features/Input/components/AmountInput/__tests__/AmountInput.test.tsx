import {AmountInput} from '../..'

import {fireEvent, render, waitFor} from '@utils'

describe('Amount Input -> Custom Input', () => {
  const testId = {
    amountTestId: 'amount-test-id',
    currencyTestId: 'currency-test-id',
    iconTestId: 'currency-icon-test-id',
  }
  it('amount input ilk render anında snapshot ile eşleşmeli', () => {
    const onChangeTextMock = jest.fn()
    const renderedInput = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    expect(renderedInput).toMatchSnapshot()
  })

  it('amount input amount ve currency inputlari ve icon var olmali', () => {
    const {amountTestId, currencyTestId, iconTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const amountInput = getByTestId(amountTestId)
    const currencyInput = getByTestId(currencyTestId)
    const iconElement = getByTestId(iconTestId)

    expect(amountInput).toBeOnTheScreen()
    expect(currencyInput).toBeOnTheScreen()
    expect(iconElement).toBeOnTheScreen()
  })

  it('amount input sadece numeric karakterleri kabul etmeli', () => {
    const {amountTestId, currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId, rerender} = render(
      <AmountInput name='test' onChangeText={onChangeTextMock} />
    )

    const amountInput = getByTestId(amountTestId)
    const currencyInput = getByTestId(currencyTestId)

    fireEvent.changeText(amountInput, 'aaA*a123')
    fireEvent.changeText(currencyInput, '1aaA*a123')

    rerender(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    expect(amountInput).toHaveProp('value', '123')
    expect(currencyInput).toHaveProp('value', '11')
  })

  it('amount input formata uygun olmalı', () => {
    const {amountTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const amountInput = getByTestId(amountTestId)
    fireEvent.changeText(amountInput, '12')

    expect(onChangeTextMock).toHaveBeenCalledWith('12.00')
  })

  it('amount inputun küsüratlı tarafı değiştiği zaman boş bir değer bırakılırsa küsürat 00 olmalı', () => {
    const {currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const currencyElement = getByTestId(currencyTestId)
    fireEvent.changeText(currencyElement, '50')
    expect(currencyElement).toHaveProp('value', '50')

    fireEvent.changeText(currencyElement, '')
    fireEvent(currencyElement, 'blur')
    expect(currencyElement).toHaveProp('value', '00')
  })

  it('amount inputta küsürat 0 iken tıklandığı zaman değer 00 dan boşa çekilmeli', () => {
    const {currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const currencyElement = getByTestId(currencyTestId)
    fireEvent.changeText(currencyElement, '00')
    fireEvent(currencyElement, 'focus')

    expect(currencyElement).toHaveProp('value', '')
  })

  it('amount inputta küsürat 0 dan büyükken tıklandığı zaman değer ne ise kalmalı', async () => {
    const {currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const currencyElement = getByTestId(currencyTestId)
    await waitFor(() => {
      fireEvent.changeText(currencyElement, '50')
    })
    fireEvent(currencyElement, 'focus')

    expect(currencyElement).toHaveProp('value', '50')
  })

  it('input render olduğu zaman klavye olarak number-pad ekranda görülmeli', () => {
    const {amountTestId, currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)

    const amountInput = getByTestId(amountTestId)
    const currencyInput = getByTestId(currencyTestId)

    expect(amountInput).toHaveProp('keyboardType', 'number-pad')
    expect(currencyInput).toHaveProp('keyboardType', 'number-pad')
  })

  it('amount inputa verilen currencyType a göre icon görülmeli', () => {
    const currencyType = 'USD'
    const onChangeTextMock = jest.fn()

    render(<AmountInput name='test' currencyType={currencyType} onChangeText={onChangeTextMock} />)

    expect(onChangeTextMock).toHaveBeenCalled()
  })

  it('kullanıcı 0.01 değeri girdiğinde doğru bir şekilde formatlanmalı', () => {
    // Arrange - Hazırlık
    const {amountTestId, currencyTestId} = testId
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<AmountInput name='test' onChangeText={onChangeTextMock} />)
    const amountInput = getByTestId(amountTestId)
    const currencyInput = getByTestId(currencyTestId)

    // Act 1 - Eylem 1: Amount alanına "0" girişi
    fireEvent.changeText(amountInput, '0')

    // Assert 1 - Doğrulama 1: "0.00" değeri kontrol edilir
    expect(onChangeTextMock).toHaveBeenCalledWith('0.00')

    // Arrange 2 - Hazırlık 2: Yeni değişikliği test etmek için mock temizlenir
    onChangeTextMock.mockClear()

    // Act 2 - Eylem 2: Currency alanına "01" girişi
    fireEvent.changeText(currencyInput, '01')

    // Assert 2 - Doğrulama 2: Leading zero korunarak "0.01" değeri doğru oluşturulmalı
    expect(onChangeTextMock).toHaveBeenCalledWith('0.01')
  })
})
