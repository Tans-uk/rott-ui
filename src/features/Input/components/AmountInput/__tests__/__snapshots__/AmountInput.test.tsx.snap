// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount Input -> Custom Input amount input ilk render anında snapshot ile eşleşmeli 1`] = `
<KeyboardProvider
  enabled={true}
>
  <View
    flex={1}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      onTouchStart={[Function]}
      relative={true}
      style={
        {
          "flex": undefined,
          "flexDirection": "column",
          "position": "relative",
        }
      }
    >
      <View
        name="test"
        style={
          {
            "flex": undefined,
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          editable={true}
          keyboardType="number-pad"
          maxLength={11}
          nativeID="amount-native-id"
          onChangeText={[Function]}
          placeholder="0"
          placeholderTextColor="rgba(161, 173, 175, 1)"
          style={
            {
              "color": "rgba(161, 173, 175, 1)",
              "fontFamily": "Markpro-Medium",
              "fontSize": 18,
              "height": 89,
              "letterSpacing": 0.5,
              "width": "auto",
            }
          }
          testID="amount-test-id"
          value=""
        />
        <View
          justifyContentFlexEnd={true}
          paddingBottom={19}
          style={
            {
              "flex": undefined,
              "flexDirection": "column",
              "justifyContent": "flex-end",
              "paddingBottom": 37,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgba(161, 173, 175, 1)",
                "fontFamily": "Markpro-Bold",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "auto",
              }
            }
          >
            ,
          </Text>
        </View>
        <TextInput
          editable={true}
          keyboardType="number-pad"
          maxLength={2}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="00"
          placeholderTextColor="rgba(161, 173, 175, 1)"
          style={
            {
              "color": "rgba(161, 173, 175, 1)",
              "fontFamily": "Markpro-Medium",
              "fontSize": 18,
              "height": 89,
              "letterSpacing": 0.5,
              "width": "auto",
            }
          }
          testID="currency-test-id"
          value=""
        />
      </View>
      <View
        absolute={true}
        bottom={16}
        right={0}
        style={
          {
            "bottom": 31,
            "flex": undefined,
            "flexDirection": "column",
            "position": "absolute",
            "right": 0,
          }
        }
      >
        <View
          style={
            {
              "flex": undefined,
              "flexDirection": "column",
            }
          }
          testID="currency-icon-test-id"
        >
          <Text
            fill="rgba(161, 173, 175, 1)"
            height={46}
            strokeWidth={0}
            testID="mock-icon"
            width={46}
          >
            MockIcon
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "flexDirection": "column-reverse",
            "justifyContent": "flex-start",
            "top": 30,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "bottom": 40,
            "flexDirection": "column",
            "justifyContent": "flex-end",
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
  </View>
</KeyboardProvider>
`;
